File: src/CashTrees_iOS/CashTrees_iOS/CashTrees_iOS/CashTreesWebConsoleRelay.swift
//
//  CashTreesWebConsoleRelay.swift
//  CashTrees_iOS
//
//  Forwards PWA `console.*` output to Console.app via os.Logger (Unified Logging).
//

import Foundation
import os
import WebKit

enum CashTreesWebConsoleRelay {
    static let handlerName = "CashTreesWebConsole"
    static let bridgeHandlerName = "CashTreesIOS"
    /// Console.app: Process CashTrees_iOS, subsystem=com.beamio.beamio, category=PWA-JS.
    static let logSubsystem = "com.beamio.beamio"
    static let logCategory = "PWA-JS"
    static let logTag = "PWA-JS"

    private static let logger = Logger(subsystem: logSubsystem, category: logCategory)

    static func logNative(_ message: String) {
        logger.info("[native] \(message, privacy: .public)")
        debugNSLog("[native] \(message)")
    }

    static func logAppBoot() {
        logNative("CashTrees_iOS app boot — filter Console.app: subsystem=\(logSubsystem) category=\(logCategory)")
    }

    static func pageUserScript(
        source: String,
        injectionTime: WKUserScriptInjectionTime,
        forMainFrameOnly: Bool
    ) -> WKUserScript {
        WKUserScript(
            source: source,
            injectionTime: injectionTime,
            forMainFrameOnly: forMainFrameOnly,
            in: .page
        )
    }

    static func injectionScript() -> String {
        """
        (function(){
          if(window.__CT_CONSOLE_PROXY_V4__)return;
          window.__CT_CONSOLE_PROXY_V4__=true;
          var CONSOLE_H='\(handlerName)';
          var BRIDGE_H='\(bridgeHandlerName)';
          var LEVELS=['log','warn','error','info','debug'];
          var pendingQueue=[];
          function stringify(v){
            if(v===undefined)return 'undefined';
            if(v===null)return 'null';
            if(typeof v==='string')return v;
            if(typeof v==='number'||typeof v==='boolean')return String(v);
            try{return JSON.stringify(v);}catch(e){
              try{return String(v);}catch(_){return '[object]';}
            }
          }
          function hasHandler(name){
            return !!(window.webkit&&window.webkit.messageHandlers&&window.webkit.messageHandlers[name]);
          }
          function sendNative(level,message){
            try{
              if(hasHandler(CONSOLE_H)){
                window.webkit.messageHandlers[CONSOLE_H].postMessage({level:level,message:message});
                return true;
              }
              if(hasHandler(BRIDGE_H)){
                window.webkit.messageHandlers[BRIDGE_H].postMessage({
                  action:'consoleLog',
                  level:level,
                  message:message
                });
                return true;
              }
            }catch(e){}
            return false;
          }
          function relay(level,args){
            var text=Array.prototype.slice.call(args).map(stringify).join(' ');
            if(!sendNative(level,text)){
              pendingQueue.push({level:level,message:text});
            }
          }
          function flushPending(){
            if(!hasHandler(CONSOLE_H)&&!hasHandler(BRIDGE_H))return false;
            while(pendingQueue.length){
              var item=pendingQueue.shift();
              sendNative(item.level,item.message);
            }
            if(!window.__CT_CONSOLE_RELAY_READY_SENT__){
              window.__CT_CONSOLE_RELAY_READY_SENT__=true;
              sendNative('info','__console_relay_ready__');
            }
            return true;
          }
          var targetConsole=window.console;
          window.console=new Proxy(targetConsole,{
            get:function(obj,prop){
              var value=Reflect.get(obj,prop);
              if(typeof prop==='string'&&LEVELS.indexOf(prop)>=0&&typeof value==='function'){
                return function(){
                  relay(prop,arguments);
                  return value.apply(obj,arguments);
                };
              }
              return typeof value==='function'?value.bind(obj):value;
            }
          });
          if(!flushPending()){
            var attempts=0;
            var timer=setInterval(function(){
              if(flushPending()||++attempts>=500)clearInterval(timer);
            },10);
          }
          if(!window.__CT_CONSOLE_ERROR_HOOKED__){
            window.__CT_CONSOLE_ERROR_HOOKED__=true;
            window.addEventListener('error',function(ev){
              try{
                var msg=(ev&&ev.message)||'script error';
                if(ev&&ev.filename){
                  msg += ' @ ' + ev.filename + ':' + (ev.lineno||0);
                }
                relay('error',[msg]);
              }catch(e){}
            });
          }
        })();
        """
    }

    static func probeScript() -> String {
        "console.log('[PWA] __pwa_console_probe__');"
    }

    static func reinject(into webView: WKWebView, reason: String) {
        let install = injectionScript()
        webView.callAsyncJavaScript(install, arguments: [:], in: nil, in: .page) { result in
            if case .failure(let error) = result {
                logNative("reinject(\(reason)) install failed: \(error.localizedDescription)")
                return
            }
            logNative("reinject(\(reason)) install ok")
            webView.callAsyncJavaScript(probeScript(), arguments: [:], in: nil, in: .page) { probeResult in
                if case .failure(let error) = probeResult {
                    logNative("reinject(\(reason)) probe failed: \(error.localizedDescription)")
                }
            }
        }
    }

    static func handle(_ message: WKScriptMessage) {
        guard message.name == handlerName else { return }
        if let body = message.body as? [String: Any] {
            emit(level: body["level"], message: body["message"])
            return
        }
        if let text = message.body as? String {
            emit(level: "log", message: text)
            return
        }
        emit(level: "log", message: String(describing: message.body))
    }

    static func handleBridgeConsoleLog(_ body: [String: Any]) {
        emit(level: body["level"], message: body["message"])
    }

    private static func emit(level: Any?, message: Any?) {
        let levelText = (level as? String)?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let resolvedLevel = (levelText?.isEmpty == false) ? levelText! : "log"
        let text: String
        if let s = message as? String {
            text = s
        } else {
            text = String(describing: message ?? "")
        }
        logPwa(level: resolvedLevel, message: text)
    }

    private static func logPwa(level: String, message: String) {
        switch level {
        case "debug":
            logger.debug("\(message, privacy: .public)")
        case "info":
            logger.info("\(message, privacy: .public)")
        case "warn", "warning":
            logger.warning("\(message, privacy: .public)")
        case "error":
            logger.error("\(message, privacy: .public)")
        case "log":
            logger.info("\(message, privacy: .public)")
        default:
            logger.info("\(message, privacy: .public)")
        }
        debugNSLog("[\(level.uppercased())] \(message)")
    }

    #if DEBUG
    private static func debugNSLog(_ message: String) {
        NSLog("[\(logTag)] %@", message)
    }
    #else
    private static func debugNSLog(_ message: String) {}
    #endif
}

/// Dedicated handler object so WKUserContentController retains a stable target.
final class CashTreesWebConsoleMessageHandler: NSObject, WKScriptMessageHandler {
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        CashTreesWebConsoleRelay.handle(message)
    }
}

File: src/CaehTrees/app/src/main/java/com/beamio/app/CashTreesWebConsoleRelay.kt
package com.beamio.app

import android.os.SystemClock
import android.util.Log
import android.webkit.ConsoleMessage
import android.webkit.JavascriptInterface
import android.webkit.WebView
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature

/**
 * Forwards PWA `console.*` / bridge `debugLog` to logcat.
 * Mirrors iOS [CashTreesWebConsoleRelay.swift].
 */
object CashTreesWebConsoleRelay {
    const val BRIDGE_NAME = "CashTreesWebConsole"
    private const val LOG_TAG = "PWA-JS"
    private const val INJECTION_FLAG = "__CT_CONSOLE_PROXY_V4__"
    private val tagConsoleRelayRegistered = "ct_console_relay_registered".hashCode()

    /** Dedup: JS proxy relays then calls original console → [onConsoleMessage] too. */
    @Volatile
    private var lastBridgeLogText: String? = null

    @Volatile
    private var lastBridgeLogAtMs: Long = 0L

    fun logNative(message: String) {
        Log.i(LOG_TAG, "[native] $message")
    }

    fun logAppBoot() {
        logNative("CashTrees Android app boot — filter logcat tag=$LOG_TAG")
    }

    fun injectionScript(): String = """
        (function(){
          if(window.$INJECTION_FLAG)return;
          window.$INJECTION_FLAG=true;
          var BRIDGE='CashTreesAndroid';
          var CONSOLE_H='$BRIDGE_NAME';
          var LEVELS=['log','warn','error','info','debug'];
          var pendingQueue=[];
          function stringify(v){
            if(v===undefined)return 'undefined';
            if(v===null)return 'null';
            if(typeof v==='string')return v;
            if(typeof v==='number'||typeof v==='boolean')return String(v);
            try{return JSON.stringify(v);}catch(e){
              try{return String(v);}catch(_){return '[object]';}
            }
          }
          function hasBridge(){
            return !!(window[BRIDGE]&&typeof window[BRIDGE].debugLog==='function');
          }
          function hasConsoleBridge(){
            return !!(window[CONSOLE_H]&&typeof window[CONSOLE_H].relay==='function');
          }
          function sendNative(level,message){
            try{
              if(hasConsoleBridge()){
                window[CONSOLE_H].relay(level,message);
                return true;
              }
              if(hasBridge()){
                window[BRIDGE].debugLog(level,message);
                return true;
              }
            }catch(e){}
            return false;
          }
          function relay(level,args){
            var text=Array.prototype.slice.call(args).map(stringify).join(' ');
            if(!sendNative(level,text)){
              pendingQueue.push({level:level,message:text});
            }
          }
          function flushPending(){
            if(!hasConsoleBridge()&&!hasBridge())return false;
            while(pendingQueue.length){
              var item=pendingQueue.shift();
              sendNative(item.level,item.message);
            }
            if(!window.__CT_CONSOLE_RELAY_READY_SENT__){
              window.__CT_CONSOLE_RELAY_READY_SENT__=true;
              sendNative('info','__console_relay_ready__');
            }
            return true;
          }
          var targetConsole=window.console;
          window.console=new Proxy(targetConsole,{
            get:function(obj,prop){
              var value=Reflect.get(obj,prop);
              if(typeof prop==='string'&&LEVELS.indexOf(prop)>=0&&typeof value==='function'){
                return function(){
                  relay(prop,arguments);
                  return value.apply(obj,arguments);
                };
              }
              return typeof value==='function'?value.bind(obj):value;
            }
          });
          if(!flushPending()){
            var attempts=0;
            (function tick(){
              if(flushPending()||++attempts>=500)return;
              setTimeout(tick,10);
            })();
          }
          if(!window.__CT_CONSOLE_ERROR_HOOKED__){
            window.__CT_CONSOLE_ERROR_HOOKED__=true;
            window.addEventListener('error',function(ev){
              try{
                var msg=(ev&&ev.message)||'script error';
                if(ev&&ev.filename){
                  msg += ' @ ' + ev.filename + ':' + (ev.lineno||0);
                }
                relay('error',[msg]);
              }catch(e){}
            });
            window.addEventListener('unhandledrejection',function(ev){
              var reason=ev.reason;
              relay('error',['Unhandled rejection: '+stringify(reason)]);
            });
          }
        })();
    """.trimIndent()

    fun probeScript(): String = "console.log('[PWA] __pwa_console_probe__');"

    /**
     * iOS parity: inject at document start (all frames) before app bundle runs.
     * Falls back to [reinject] on page lifecycle when feature unsupported.
     */
    fun registerDocumentStartScript(webView: WebView) {
        if (webView.getTag(tagConsoleRelayRegistered) == true) return
        if (!WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
            logNative("DOCUMENT_START_SCRIPT unsupported — using onPageStarted reinject only")
            webView.setTag(tagConsoleRelayRegistered, true)
            return
        }
        WebViewCompat.addDocumentStartJavaScript(
            webView,
            injectionScript(),
            setOf("*"),
        )
        webView.setTag(tagConsoleRelayRegistered, true)
        logNative("document-start console relay registered (all origins)")
    }

    fun reinject(webView: WebView, reason: String) {
        webView.evaluateJavascript(injectionScript(), null)
        webView.evaluateJavascript(probeScript(), null)
        logNative("console relay reinjected ($reason)")
    }

    fun handleBridgeConsoleLog(level: String, message: String) {
        rememberBridgeLog(message)
        writeLog(level, message)
    }

    /** [WebChromeClient.onConsoleMessage] fallback (subframes / early boot). */
    fun handleWebViewConsoleMessage(
        line: String,
        level: ConsoleMessage.MessageLevel,
        source: String?,
        lineNumber: Int,
    ) {
        val trimmed = line.trim()
        if (trimmed.isEmpty()) return
        if (isDuplicateOfRecentBridgeLog(trimmed)) return

        val suffix = buildString {
            if (!source.isNullOrBlank()) {
                append(" @ ")
                append(source)
                if (lineNumber > 0) append(':').append(lineNumber)
            }
        }
        val levelName = when (level) {
            ConsoleMessage.MessageLevel.ERROR -> "error"
            ConsoleMessage.MessageLevel.WARNING -> "warn"
            ConsoleMessage.MessageLevel.DEBUG -> "debug"
            ConsoleMessage.MessageLevel.TIP -> "info"
            ConsoleMessage.MessageLevel.LOG -> "log"
            else -> "log"
        }
        writeLog(levelName, trimmed + suffix)
    }

    private fun rememberBridgeLog(message: String) {
        lastBridgeLogText = message
        lastBridgeLogAtMs = SystemClock.elapsedRealtime()
    }

    private fun isDuplicateOfRecentBridgeLog(message: String): Boolean {
        val prev = lastBridgeLogText ?: return false
        if (prev != message) return false
        return SystemClock.elapsedRealtime() - lastBridgeLogAtMs < 120L
    }

    private fun writeLog(level: String, message: String) {
        val text = message.ifEmpty { "(empty)" }
        when (level.lowercase()) {
            "error" -> Log.e(LOG_TAG, text)
            "warn", "warning" -> Log.w(LOG_TAG, text)
            "info" -> Log.i(LOG_TAG, text)
            "debug" -> Log.d(LOG_TAG, text)
            "log" -> Log.i(LOG_TAG, text)
            else -> Log.i(LOG_TAG, text)
        }
    }

    /** Secondary JS interface: `window.CashTreesWebConsole.relay(level, message)`. */
    class JsRelay {
        @JavascriptInterface
        fun relay(level: String, message: String) {
            handleBridgeConsoleLog(level, message)
        }
    }
}
